backup-s3/src/test/scala/io/aiven/guardian/kafka/backup/s3/RealS3BackupClientSpec.scala

```diff
@@ -16,4 +56,289 @@ class RealS3BackupClientSpec extends AnyPropTestKit(ActorSystem("RealS3BackupCli
+
+      implicit val config: S3Config     = s3Config
+      implicit val backupConfig: Backup = Backup(PeriodFromFirst(1 minute))
+
+      val producerSettings = createProducer()
+
+      val backupClient =
+        new BackupClient(Some(s3Settings))(new KafkaClient(configureConsumer = baseKafkaConfig),
+                                           implicitly,
+                                           implicitly,
+                                           implicitly,
+                                           implicitly
+        )
+
+      val adminClient = AdminClient.create(
+        Map[String, AnyRef](
+          CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG -> container.bootstrapServers
+        ).asJava
+      )
+
+      val createTopics = adminClient.createTopics(topics.map { topic =>
+        new NewTopic(topic, 1, 1.toShort)
+      }.asJava)
+
+      val calculatedFuture = for {
+        _ <- createTopics.all().toCompletableFuture.asScala
+        _ <- createBucket(s3Config.dataBucket)
+        _ = backupClient.backup.run()
+        _ <- akka.pattern.after(KafkaInitializationTimeoutConstant)(
+               baseSource
+                 .runWith(Producer.plainSink(producerSettings))
+             )
+        _   <- sendTopicAfterTimePeriod(1 minute, producerSettings, topics.head)
+        key <- getKeyFromSingleDownload(s3Config.dataBucket)
+        downloaded <-
+          S3.download(s3Config.dataBucket, key)
+            .withAttributes(s3Attrs)
+            .runWith(Sink.head)
+            .flatMap {
+              case Some((downloadSource, _)) =>
+                downloadSource.via(CirceStreamSupport.decode[List[Option[ReducedConsumerRecord]]]).runWith(Sink.seq)
+              case None => throw new Exception(s"Expected object in bucket ${s3Config.dataBucket} with key $key")
+            }
+
+      } yield downloaded.toList.flatten.collect { case Some(reducedConsumerRecord) =>
+        reducedConsumerRecord
+      }
+
+      val downloaded = calculatedFuture.futureValue
+
+      val downloadedGroupedAsKey = downloaded
+        .groupBy(_.key)
+        .view
+        .mapValues { reducedConsumerRecords =>
+          reducedConsumerRecords.map(_.value)
+        }
+        .toMap
+
+      val inputAsKey = data
+        .groupBy(_.key)
+        .view
+        .mapValues { reducedConsumerRecords =>
+          reducedConsumerRecords.map(_.value)
+        }
+        .toMap
+
+      downloadedGroupedAsKey must matchTo(inputAsKey)
+    }
+  }
+
+  def waitUntilBackupClientHasCommitted(backupClient: BackupClientChunkState[_],
+                                        step: FiniteDuration = 100 millis,
+                                        delay: FiniteDuration = 5 seconds
+  ): Future[Unit] =
+    if (backupClient.processedChunks.size() > 0)
+      akka.pattern.after(delay)(Future.successful(()))
+    else
+      akka.pattern.after(step)(waitUntilBackupClientHasCommitted(backupClient, step, delay))
```

Comment:
Nit pick, maybe move this auxiliary function on top, so all properties are together one after the other

Reply:
Done